src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -175,4 +159,184 @@ export class MagicWandService {
+   */
+  getIsScribbleMask(scribbles: Set<number>, imgData: ImageData,
+      pixelCoord: Array<number>, tolerance: number): boolean {
+    const curX: number = pixelCoord[0];
+    const curY: number = pixelCoord[1];
+
+    // Gets array of color attributes of current pixel.
+    const curPixel: Array<number> = this.dataArrayToRgb(imgData, curX, curY);
+    // Work with tolerance logic in squared space for Euclidean distance.
+    const squaredTolerance = tolerance * tolerance;
+
+    for (let pixelIndex of scribbles) {
+      const refPixelCoord: Array<number> = 
+          this.pixelIndexToXYCoord(pixelIndex, imgData.width);
+      const x = refPixelCoord[0];
+      const y = refPixelCoord[1];
+      const refPixel = this.dataArrayToRgb(imgData, x, y);
+
+      const colorDifference = this.rgbEuclideanDist(refPixel, curPixel);
+
+      // If the curPixel is tolerable for at least 1 of the reference pixels,
+      // then the curPixel will be part of the mask.
+      if (colorDifference <= squaredTolerance) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**@returns {Set<number>} filtered set of pixels by removing indices 
+   * of pixels whose color are too similar to the original pixel. 
+   * Judgement of 'too similar' is decided by a factor of 
+   * @param {number} tolerance and color value itself is retrieved from
+   * @param {ImageData} imgData
+  */
+  filterScribbles(scribbles: Set<number>, originalPixel: Array<number>, 
+      imgData: ImageData, tolerance: number): Set<number> {
+    let result: Set<number> = new Set();
+
+    // Work with tolerance logic in squared space for Euclidean distance.
+    const squaredTolerance = tolerance * tolerance;
+
+    for (let pixelIndex of scribbles) {
+      const pixelCoord: Array<number> = 
+          this.pixelIndexToXYCoord(pixelIndex, imgData.width);
+      const x = pixelCoord[0];
+      const y = pixelCoord[1];
+      const curPixel = this.dataArrayToRgb(imgData, x, y);
+
+      const colorDifference = this.rgbEuclideanDist(originalPixel, curPixel);
+
+      if (colorDifference !== 0 && colorDifference > squaredTolerance / 2) {
+        result.add(pixelIndex);
+      }
+    }
+
+    return result;
+  }
+
+  /**@returns {Array<number> [x, y]} a 2-D coordinate by converting 
+   * @param {number} pixelIndex into the coordsponding [x, y] coordinate.
+   * @param {number} width should be the pixel width of the image that the 
+   * pixelIndex belongs to.
+   */
+  pixelIndexToXYCoord(pixelIndex: number, width: number): Array<number> {
+    return [((pixelIndex / 4) % width), ((pixelIndex / 4) / width)];
+  }
+
+  /**Does the general floodfill algorithm, and switches to 
+   * scribbleFloodfill() if 
+   * @param {Set<number>} scribbles is provided as a non-empty set.
+   */
+  doFloodfill(imgData: ImageData, xCoord: number, yCoord: number,
+      tolerance: number, scribbles: Set<number>): Set<number> {
```

Comment:
You don't need to do this in this PR, but something to consider for future clean ups.  You have a mixed representation of pixel coordinates.  They are either (x,y) or pixel indices.  You could consider picking one and using it throughout all headers.  By making this consistent it makes it easier to work with and to know what to pass in.

Reply:
Thanks! I'll note this down.